File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

void check(void);

int main(void)
{
    check();
}

void check(void)
{
    // Finds the number of digits in the user's entered number and assigns it to numDigs
    int numDigs = 0;
    long n = 0;
    n = get_long("Number: ");
    long copy = n;
    while (copy > 0)
    {
        copy /= 10;
        numDigs++;
    }

    // Prints invalid if the user's number isn't a valid credit card number length
    if (numDigs != 13 && numDigs != 15 && numDigs != 16)
    {
        printf("INVALID\n");
        return;
    }

    int normalSum = 0;
    int digSum = 0;

    // Multiplies every other number by 2 and finds the sum of the digits
    // Also finds the normal sum of the numbers that weren't multiplied by 2
    for (int i = numDigs; i > 0; i--)
    {
        long temp = 1;
        for (int j = 1; j < i; j++)
        {
            temp *= 10;
        }

        if (i % 2 == 1)
        {
            normalSum += (n / temp) % 10;
        }
        else
        {
            int num = ((n / temp) % 10) * 2;
            digSum += num % 10 + num / 10;
        }
    }

    // Checks to see if the card is a valid credit card number
    // Prints the associated company for the card if the card is valid
    if ((digSum + normalSum) % 10 == 0)
    {
        int firstTwo = n / (divNumber / 100);
        if ((firstTwo == 34 || firstTwo == 37) && (numDigs == 15))
        {
            printf("AMEX\n");
            return;
        }
        else if ((firstTwo == 51 || firstTwo == 52 || firstTwo == 53 || firstTwo == 54 ||
                  firstTwo == 55) &&
                 (numDigs == 16))
        {
            printf("MASTERCARD\n");
            return;
        }
        else if (firstTwo / 10 == 4 && (numDigs == 13 || numDigs == 16))
        {
            printf("VISA\n");
            return;
        }
    }

    // Prints invalid for when the number is technically valid, but doesn't match with a company
    printf("INVALID\n");
    return;
}

File: pset6/sentimental-credit/credit.py
from cs50 import get_string

ans = get_string("Number: ")
if len(ans) == 13 or len(ans) == 15 or len(ans) == 16:
    pass
else:
    print("INVALID")
    exit()
try:
    int(ans)
except ValueError:
    print("INVALID")
    exit()

counter = 0
digitSum = 0
normalSum = 0
for i in ans[::-1]:
    dig = int(ans[len(ans) - counter - 1: len(ans) - counter])
    if (counter % 2 == 1):
        digitSum += (dig * 2) % 10
        if (dig * 2) / 10 >= 1:
            digitSum += int((dig * 2) / 10)
    else:
        normalSum += dig
    counter += 1

if (digitSum + normalSum) % 10 == 0:
    if len(ans) == 15 and int(ans[0: 2]) in {34, 37}:
        print("AMEX")
        exit()
    elif len(ans) == 16 and int(ans[0: 2]) in {51, 52, 53, 54, 55}:
        print("MASTERCARD")
        exit()
    elif len(ans) in {13, 16} and int(ans[0: 1]) == 4:
        print("VISA")
        exit()
    else:
        print("INVALID")
        exit()

print("INVALID")
exit()
